salt/states/boto_ec2.py

```diff
@@ -452,3 +452,151 @@ def snapshot_created(name, ami_name, instance_name, wait_until_available=True, w
+
+    for k, v in attributes.iteritems():
+        curr = __salt__['boto_ec2.get_attribute'](k, instance_id=instance_id, region=region, key=key,
+                                                  keyid=keyid, profile=profile)
+        if curr and curr.get(k) == v:
+            continue
+        else:
+            try:
+                r = __salt__['boto_ec2.set_attribute'](attribute=k, attribute_value=v,
+                                                       instance_id=instance_id, region=region,
+                                                       key=key, keyid=keyid, profile=profile)
+            except SaltInvocationError as e:
+                ret['result'] = False
+                ret['comment'] = 'Failed to set attribute {0} to {1} on instance {2}.'.format(k, v, instance_name)
+                return ret
+            ret['changes'] = ret['changes'] if ret['changes'] else {'old': {}, 'new': {}}
+            ret['changes']['old'][k] = curr.get(k)
+            ret['changes']['new'][k] = v
+
+    return ret
+
+
+def instance_absent(name, instance_name=None, instance_id=None,
+                     region=None, key=None, keyid=None, profile=None):
+    '''
+    Ensure an EC2 instance does not exist (is stopped and removed).
+    '''
+    ret = {'name': name,
+           'result': True,
+           'comment': '',
+           'changes': {}
+          }
+    running_states = ('pending', 'rebooting', 'running', 'stopping', 'stopped')
+
+    if not instance_id:
+        try:
+            instance_id = __salt__['boto_ec2.get_id'](name=instance_name if instance_name else name,
+                                                      region=region, key=key, keyid=keyid,
+                                                      profile=profile, in_states=running_states)
+        except CommandExecutionError as e:
+            ret['result'] = None
+            ret['comment'] = 'Couldn\'t determine current status of instance {0}.'.format(instance_name)
+            return ret
+
+    exists = __salt__['boto_ec2.exists'](instance_id=instance_id, region=region,
+                                         key=key, keyid=keyid, profile=profile)
+    if not exists:
+        ret['result'] = True
+        ret['comment'] = 'Instance {0} is already gone.'.format(instance_id)
+        return ret
+
+    ### Honor 'disableApiTermination' - if you want to override it, first use set_attribute() to turn it off
+    no_can_do = __salt__['boto_ec2.get_attribute']('disableApiTermination', instance_id=instance_id,
+                                                   region=region, key=key, keyid=keyid, profile=profile)
+    if no_can_do.get('disableApiTermination') == True:
+        ret['result'] = False
+        ret['comment'] = 'Termination of instance {0} via the API is disabled.'.format(instance_id)
+        return ret
+
+    r = __salt__['boto_ec2.terminate'](instance_id=instance_id, name=instance_name, region=region,
```

Comment:
OUCH, you're right.  Fixing now...  I've been "testing" all this time by
building and deleting a VPC repeatedly, and wasn't thinking about test mode
at all...

On Tue, Oct 6, 2015 at 6:38 PM, Ryan Lane <notifications@github.com> wrote:

> In salt/states/boto_ec2.py
> <https://github.com/saltstack/salt/pull/27728#discussion_r41331874>:
>
> > +    exists = __salt__['boto_ec2.exists'](instance_id=instance_id, region=region,
> > +                                         key=key, keyid=keyid, profile=profile)
> > +    if not exists:
> > +        ret['result'] = True
> > +        ret['comment'] = 'Instance {0} is already gone.'.format(instance_id)
> > +        return ret
> > +
> > +    ### Honor 'disableApiTermination' - if you want to override it, first use set_attribute() to turn it off
> > +    no_can_do = __salt__['boto_ec2.get_attribute']('disableApiTermination', instance_id=instance_id,
> > +                                                   region=region, key=key, keyid=keyid, profile=profile)
> > +    if no_can_do.get('disableApiTermination') == True:
> > +        ret['result'] = False
> > +        ret['comment'] = 'Termination of instance {0} via the API is disabled.'.format(instance_id)
> > +        return ret
> > +
> > +    r = __salt__['boto_ec2.terminate'](instance_id=instance_id, name=instance_name, region=region,
>
> Also missing test mode.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/saltstack/salt/pull/27728/files#r41331874>.
>